src/python/pants/backend/jvm/tasks/coverage/engine.py

```diff
@@ -5,63 +5,80 @@
 from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
                         unicode_literals, with_statement)
 
-from abc import abstractmethod, abstractproperty
+from abc import abstractmethod
 
 from pants.util.meta import AbstractClass
+from pants.util.objects import datatype
 
 
 class CoverageEngine(AbstractClass):
-  """The interface JVM code coverage processors must support."""
+  """The interface JVM code coverage processors must support.
 
-  @abstractmethod
-  def instrument(self):
-    """Instruments JVM bytecode for coverage tracking."""
+  This is a stateful interface that will be called via the following sequence:
+
+  1. instrument
+  2. run_modifications
+  3. report
+  4. maybe_open_report
+
+  In this sequence, the `run_modifications` call in step 2 may occur multiple times if coverage data
+  is being collected in parts. The `report` call in step 3 should be able to merge the results of
+  these multiple runs.
+  """
+
+  class RunModifications(datatype('RunModifications', ['classpath_prepend', 'extra_jvm_options'])):
+    """Modifications that should be made to the java command where code coverage is collected.
+
+    The `classpath_prepend` field should be an iterable of classpath elements to prepend to java
+    command classpath.
+
+    The `extra_jvm_options` should be a list of jvm options to include when executing the java
+    command.
+    """
+
+    @classmethod
+    def create(cls, classpath_prepend=None, extra_jvm_options=None):
+      return cls(classpath_prepend=classpath_prepend or (),
+                 extra_jvm_options=extra_jvm_options or [])
```

Comment:
Any reason not to make both of these tuples?

Reply:
Not a super great reason, but docs align and the one usage expects the docs (a list vs a tuple) - it adds this list to another in a list1 + thisList + otherList sandwich.